### ISSUE TITLE
logarithm with base 0 should raise an error

### ISSUE THREAD

**Issue by DaveWitteMorris** [sagemath/sage#38971]
### Expected Behavior

Logarithms with a base of 0 do not exist, so trying to calculate one should raise a `ValueError`.

### Actual Behavior

Sagemath seems to think that logarithms with a base of 0 are equal to 0:
```
sage: log(2, 0)
0
sage: log(2.0, 0)
-0.000000000000000
sage: log(x, 0)
0
```

### Environment

- **OS**: MacOS 14.6.1
- **Sage Version**: 10.5.b9

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

**Comment by user202729** [comment]
Probably because `log(2, 0) = log(2)/log(0) = log(2)/(-Infinity) = 0`.

We do have $\lim_(x → 0) \log(2)/\log(x) = 0$ anyway, so the behavior is not entirely unreasonable. Besides, the change breaks backwards compatibility.

**Comment by DaveWitteMorris** [comment]
Because of the limit interpretation, the value of 0 probably makes sense to a physicist (or engineer), but I think it is clearly incorrect from a mathematical point of view, because the definition of `log` is: `b^log(a,b) = a` and this is clearly not true when `b = 0`. (If someone wants a limit, they can take a limit.) So I think the current behaviour counts as a bug and needs to be changed (even if it breaks some code). 

If anyone thinks this may be controversial, we should ask for opinions on [sage-devel](https://groups.google.com/g/sage-devel/).

**Comment by DaveWitteMorris** [comment]
Actually, since the change could break code (and the bug has been around for a long time, so fixing it is not urgent), I think the first step would be to deprecate the use of 0 as a base.  (I expect 10.5 to be released soon, so I think this will have to wait for 10.6.)  A year later, we can make it raise an error.  Probably the deprecation should be mentioned on sage-devel, to see if there are any objections.

**Comment by user202729** [comment]
There's also the issue that testing if a symbolic expression is zero is a very nontrivial problem, but I'm not sure this is a serious roadblock here.

**Comment by DaveWitteMorris** [comment]
I don't think testing for zero will cause a problem here. The issue is about what sagemath should do when it has determined that the base is zero, so there shouldn't be any additional checks for 0.

A related issue (perhaps a separate issue): Returning the natural logarithm is even worse than returning zero, but that's what sagemath does when 0 is specified as the base via keyword:
```
sage: log(6, base=0)
log(6)
```
The problem is in line 1141 of src/sage/rings/polynomial/multi_polynomial_ring.py:
```
    base = kwds.pop('base', None)
    if base:
        args = args + (base,)
```
It should be "`if base is not None`", instead of "`if base`".

**Comment by Noel-Roemmele** [comment]
Pull Request for the sub-issue mentioned by Dave Morris has been created.

**Comment by devansh-srv** [comment]
@DaveWitteMorris  Now that Sage is on ** 10.6**  shall the log raise an error when it gets zero?

**Comment by devansh-srv** [comment]
This is the definition of log function in `src/sage/rings/integer.pyx`
**functions signature**
`    def log(self, m=None, prec=None):
`
it clearly says that if a m<=0 then it raises a ValueError
but I think rather than a ValueError
it should provide an answer to it with some value of prec using mpfr as it is doing below

```
        cdef int self_sgn
        if m is not None and m <= 0:
            raise ValueError("log base must be positive")
        self_sgn = mpz_sgn(self.value)
        if self_sgn < 0 and prec is None:
            from sage.symbolic.ring import SR
            return SR(self).log(m)
        if prec:
            if self_sgn >= 0:
                from sage.rings.real_mpfr import RealField
                return RealField(prec)(self).log(m)
            else:
                from sage.rings.complex_mpfr import ComplexField
                return ComplexField(prec)(self).log(m)

        if m is None:
            from sage.functions.log import function_log
            return function_log(self,dont_call_method_on_arg=True)
        try:
            m = Integer(m)
        except (ValueError, TypeError):
            pass

        if isinstance(m, Integer):
            elog = self.exact_log(m)
            if elog == -sage.rings.infinity.infinity or m**elog == self:
                return elog

        if isinstance(m, Rational) and m.numer() == 1:
            elog = -self.exact_log(m.denom())
            if m**elog == self:
                return elog

        from sage.functions.log import function_log
        return function_log(self, dont_call_method_on_arg=True)/\
            function_log(m, dont_call_method_on_arg=True)

```

**Comment by devansh-srv** [comment]
@user202729 
I just realised that removing the check for `m<=0` gives 
```
sage: 8.log(0)
0
sage: (8.0).log(0.0)
-0.000000000000000
sage: 8.log(-2)
3*log(2)/(I*pi + log(2))
sage: (8.0).log(-2.0)
0.139260970636224 - 0.631180872623791*I
```  
which is pretty much correct and makes log consistent
should I open a PR and patch this up?

**Comment by DaveWitteMorris** [comment]
No, the current behaviour of this method is correct (at least for 0), so your suggested change is going in the wrong direction. As stated in the description of this issue, the expected behaviour is: "Logarithms with a base of 0 do not exist, so trying to calculate one should raise a `ValueError`."

Logarithms are calculated by several different methods/functions in the sagemath code. This particular method already behaves correctly, but others do not. As mentioned in a previous comment, the ones that currently do not raise an error should be modified to produce a deprecation message to warn the user that the method will raise an error at some point in the future. 

When the various logarithm methods/functions have been located, a message should be posted on sage-devel to point out that we plan to change them.

**Comment by devansh-srv** [comment]
@DaveWitteMorris but what about the issue with -2
```
sage: 8.log(-2)
3*log(2)/(I*pi + log(2))
sage: (8.0).log(-2.0)
0.139260970636224 - 0.631180872623791*I
```
I get that the other methods included such as  that in  real_mpfr and complex_mpfr should give a deprecation warning but what happens when we use negative base

**Comment by devansh-srv** [comment]
for this the code should handle negative values and zero separately

**Comment by devansh-srv** [comment]
> This issue ([#38971](https://github.com/sagemath/sage/issues/38971)) is about logarithms whose base is 0. If you want to change the behaviour of logarithms with a negative base, then I think you should open a separate issue, because it will need to be discussed (but I doubt that I will take part in the discussion).

I get that but there is an issue with the base value being 0
```
sage: log(8,0)
0
sage: log(8.0,0.0)
-0.000000000000000
sage: 8.log(0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[3], line 1
----> 1 Integer(8).log(Integer(0))

File ~/SAGE_ROOT/src/sage/rings/integer.pyx:2861, in sage.rings.integer.Integer.log()
   2859 cdef int self_sgn
   2860 if m is not None and m <= 0:
-> 2861     raise ValueError("log base must be positive")
   2862 self_sgn = mpz_sgn(self.value)
   2863 if self_sgn < 0 and prec is None:

ValueError: log base must be positive
```
There is a discrepancy in the return value of log
as you can see 
`log(8,0)` returns 0 while `8.log(0)` returns **ValueError** and `8.0.log(0.0)` definitely return `-0.000000000000000` which is mpfr which I think is still fine  but the integer logs are definitely confusing

**Comment by DaveWitteMorris** [comment]
I do not really understand your comment, because examples like this were already discussed above, and I do not understand how your comments relate to the previous discussion.

I had said that logarithms to base 0 should always raise an exception, but, after further thought, I think it's fine for inexact calculations (such as `log(2.0,0.0)`) to return 0.0. However, I still think that exact calculations (including the symbolic calculation `log(x,0)`) should raise an error.

Different methods or functions with the same name can have different points of view, so the fact that they behave differently is not necessarily a bug. In particular, `log(8,0)` and `log(8.0,0.0)` can be expected to behave differently, because (as can be seen from the documentation) one is an exact (or symbolic) calculation, but the other is a numerical approximation. The logarithm cannot be calculated exactly, so I think the exact calculation should raise an error.
